plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/ReusableConnectionFactory.java

```diff
@@ -0,0 +1,184 @@
+import java.time.Duration;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.cache.RemovalCause.EXPLICIT;
+import static io.trino.plugin.jdbc.JdbcErrorCode.JDBC_ERROR;
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public final class ReusableConnectionFactory
+        implements ConnectionFactory, QueryEventListener
+{
+    @GuardedBy("this")
+    private final Cache<String, Connection> connections;
+    private final ConnectionFactory delegate;
+
+    @Inject
+    public ReusableConnectionFactory(@ForReusableConnectionFactory ConnectionFactory delegate)
+    {
+        this(delegate, Duration.ofSeconds(2), 10);
+    }
+
+    ReusableConnectionFactory(ConnectionFactory delegate, Duration duration, long maximumSize)
+    {
+        this.connections = createConnectionsCache(duration, maximumSize);
+        this.delegate = requireNonNull(delegate, "delegate is null");
+    }
+
+    // CacheBuilder.build(CacheLoader) is forbidden, because it does not support eviction for ongoing loads.
+    // In this class, loading is not used and cache is used more as a map. So this is safe.
+    @SuppressModernizer
+    private static Cache<String, Connection> createConnectionsCache(Duration duration, long maximumSize)
+    {
+        requireNonNull(duration, "duration is null");
+        return CacheBuilder.newBuilder()
+                .maximumSize(maximumSize)
+                .expireAfterWrite(duration)
+                .removalListener(ReusableConnectionFactory::onRemoval)
+                .build();
+    }
+
+    private static void onRemoval(RemovalNotification<String, Connection> notification)
+    {
+        if (notification.getCause() == EXPLICIT) {
+            // connection was taken from the cache
+            return;
+        }
+        try {
+            requireNonNull(notification.getValue(), "notification.getValue() is null");
+            notification.getValue().close();
+        }
+        catch (SQLException e) {
+            throw new TrinoException(JDBC_ERROR, e);
+        }
+    }
+
+    @Override
+    public Connection openConnection(ConnectorSession session)
+            throws SQLException
+    {
+        String queryId = session.getQueryId();
+        Connection connection = getConnection(session, queryId);
+        return new CachedConnection(queryId, connection);
+    }
+
+    private Connection getConnection(ConnectorSession session, String queryId)
+            throws SQLException
+    {
+        Connection connection = connections.asMap().remove(queryId);
+        if (connection != null) {
+            return connection;
+        }
+        return delegate.openConnection(session);
+    }
+
+    @Override
+    public void beginQuery(ConnectorSession session) {}
+
+    @Override
+    public void cleanupQuery(ConnectorSession session)
+    {
+        Connection connection = connections.asMap().remove(session.getQueryId());
+        if (connection != null) {
```

Comment:
Reused connection stays open as there could be next method that could try to use that connection. So this method closes it.

Are you asking when `connection == null`? Yes it may happen that no connections were opened at all. It may happen in access control error where use has no access to catalog.

Reply:
Was asking about `connection == null` - thanks for explaining when it can happen. Will help me review the test class.